Webpack (War on Globals)

btkostner opened the pull request:
I'm sorry that git has such a hard time understanding the `mv` command. Even after `git mv` and `git commit`. 

**Things that `git` sucks at detecting:**
- Files being moved from `scripts` to `_scripts`
- Adding header to all JS files so people have an idea of what they do
- Indenting most of the code because of a single line addition

**The major changes in this PR:**
- The `scripts` directory was moved to `_scripts`
- Added webpack to build script files
- Move away from external scripts served in HTML and do it in JS

**Webpack:**
- Moved away from `gulp` to `webpack` so we can bundle javascript. This means you can now use `include jQuery from 'lib/jquery'`. +3 to charisma
- Added babel for transcoding. We can now do proper classes and use a cleaner syntax. +2 to charisma
- Simplified the CSS building and testing

**What is this external script loading thing?**
So instead of having an HTML document that would load `jQuery` and `Chart.js` and a bunch of other dependencies, we load a single (hopefully) file, and that JS file is responsible for loading what it needs. This should help a lot with modularity and expansion.

**Other small changes:**
- Added a `main.js` page that loads all of those pesky snippets we have like `smooth-scrolling` and `external-links`. -3 HTTP calls
- Moved `highlight.js` to jsdelivr CDN.
- Added a `npm run start:watch` command that launches PHP server and runs `build:watch`

lewisgoddard commented:
Is the build command still `npm run build` or is webpack a separate command?

lewisgoddard commented:
Also, git seems to have understood fine https://github.com/elementary/mvp/pull/1294/commits/dc986f5d64565916b679e4554013db18a9ab22da

btkostner commented:
All the commands are the same
- `npm start` for php server 
- `npm install` for deps
- `npm run build` for building
- `npm run build:watch` for building on new things 

New command
- `npm run dev` starts php server and `build:watch`

btkostner commented:
Considering we are nearing Loki release. Please don't merge this until after.

lewisgoddard commented:
Ping #1309

btkostner commented:
Alright this should be good and ready. Sorry if I overstepped a bit while cleaning up the code. There is more changes than what I probably should have done, but on the bright side, I got a head start on #1282. 

Some of the changes:
- I added an `api/config.php` file that outputs JSON variables used throughout the site, so no more inline script variables (except on the error pages).
- I cleaned up the `_scripts` folder a bit (needs more) by stuffing all the compiled scripts to `_scripts/pages`
- Moved Google analytics to a lib just like everything else
- Cleaned up some PHP headers while I was removing script tags

I would recommend double checking and going around the pages with dev console up and making sure there are no errors

lewisgoddard commented on _scripts/lib/jquery.js:
```diff
@@ -0,0 +1,16 @@
+import Script from 'scriptjs'
+
+export default new Promise((resolve, reject) => {
+    Script('https://cdn.jsdelivr.net/jquery/3.1.0/jquery.min.js', () => {
```
Most of these are in the format `https://cdn.jsdelivr.net/g/jquery@3.1.0`

btkostner replied:
Fixed
